### PR TITLE
Fix CollectionsByProductIdLoader if collection is missing

### PR DESCRIPTION
**Pull request by Air-t**
In some cases, the collections list may contain `None` instance. Fixes `AttributeError` when that happens.

Fixes: https://linear.app/saleor/issue/SHOPX-683/attributeerror-nonetype-object-has-no-attribute-id

# Impact

- [ ] New migrations
- [ ] New/Updated API fields or mutations
- [ ] Deprecated API fields or mutations
- [ ] Removed API types, fields, or mutations

# Docs

- [ ] Link to documentation:

# Pull Request Checklist

- [ ] Privileged queries and mutations are either absent or guarded by proper permission checks
- [ ] Database queries are optimized and the number of queries is constant
- [ ] Database migrations are either absent or optimized for zero downtime
- [ ] The changes are covered by test cases
- [ ] All new fields/inputs/mutations have proper labels added (`ADDED_IN_X`, `PREVIEW_FEATURE`, etc.)
- [ ] All migrations have proper dependencies
- [ ] All indexes are added concurrently in migrations
- [ ] All RunSql and RunPython migrations have revert option defined

### CODE REVIEW

File: saleor/graphql/product/dataloaders/products.py [hunk 1]
```diff
@@ -548,7 +547,7 @@ def batch_load(self, keys):
             product_collection_map[pid].append(cid)
 
         def map_collections(collections):
-            collection_map = {c.id: c for c in collections}
+            collection_map = {c.id: c for c in collections if c is not None}
```
Comment (korycins) [inline saleor/graphql/product/dataloaders/products.py:550]: It won't solve the problem. You will change the line where an error can occur. The error is raised as probably the collection was removed in the meantime. In line 552, there is fetch via key, which will raise an exception, when the collection ID doesn't exist - and there is a possibility that the collection will not exist as you added validation for `None`.

Comment (Air-t) [reply saleor/graphql/product/dataloaders/products.py:550]: thanks, I checked the case when a Collection is removed. Now there is no issue with handling that situation.

File: saleor/graphql/product/dataloaders/products.py [hunk 2]
```diff
@@ -548,9 +547,13 @@ def batch_load(self, keys):
...
             return [
-                [collection_map[cid] for cid in product_collection_map[pid]]
+                [
+                    collection_map[cid]
+                    for cid in product_collection_map[pid]
+                    if cid in collection_map
```
Comment (fowczarek) [inline saleor/graphql/product/dataloaders/products.py:555]: @Air-t Dataloader should always return a list of the same length as the list of keys. This solution doesn't meet this requirement. I think you should start by creating a unit test with bug reproduction.

Comment (Air-t) [reply saleor/graphql/product/dataloaders/products.py]: I managed to reproduce it removing a collection while a dataloader `batch_load` was partially executed and `product_collection_pairs` list was created. Now both keys and return value lengths match. 
In a reproduced example, other functions had to be fixated to not fail while the collection was resolved as None by dataloader

Comment (fowczarek) [reply saleor/graphql/product/dataloaders/products.py]: Could you write unit tests using the `before_after` module to reproduce the described case? We use this module in some tests e.g `saleor/checkout/tests/test_checkout_complete.py::test_complete_checkout_checkout_was_deleted_before_completing`

**Review by fowczarek — APPROVED** [review]

**Review by zedzior — APPROVED** [review]

**Review by IKarbowiak — APPROVED** [review]

**Pull request merged**